File: platforms/ios/Dezel/Dezel/Core/Source.swift
import Foundation

/**
 * A context source file.
 * @class Source
 * @since 0.7.0
 */
open class Source {

	//--------------------------------------------------------------------------
	// MARK: Enum
	//--------------------------------------------------------------------------

	/**
	 * Source categories.
	 * @enum Category
	 * @since 0.7.0
	 */
	public enum `Type` {
		case style
		case script
	}

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * The source's category.
	 * @property category
	 * @since 0.7.0
	 */
	private(set) public var type: Type

	/**
	 * The source's location.
	 * @property location
	 * @since 0.7.0
	 */
	private(set) public var path: String

	/**
	 * The source's data.
	 * @property data
	 * @since 0.7.0
	 */
	public lazy var data: String = {
		return self.read()
	}()

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 * @hidden
	 */
	public init(type: Type, path: String) {
		self.type = type
		self.path = path
	}

	/**
	 * @method read
	 * @since 0.7.0
	 * @hidden
	 */
	private func read() -> String{

		do {

			if (self.path.hasPrefix("http://") ||
				self.path.hasPrefix("https://")) {
				return try String(contentsOf: URL(string: self.path)!)
			}

			return try String(contentsOfFile: Bundle.main.path(forResource: self.path, ofType: nil)!)

		} catch _ {
			fatalError("Cannot load source at location \(self.path).")
		}
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/application/Source.kt
package ca.logaritm.dezel.application

import android.content.Context
import ca.logaritm.dezel.networking.RemoteFileLoader
import java.io.IOException

/**
 * @class Source
 * @since 0.7.0
 */
open class Source(context: Context, type: Type, path: String) {

	//--------------------------------------------------------------------------
	// Enum
	//--------------------------------------------------------------------------

	/**
	 * @enum Category
	 * @since 0.7.0
	 */
	public enum class Type {
		STYLE,
		SCRIPT
	}

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property context
	 * @since 0.7.0
	 */
	public var context: Context
		private set

	/**
	 * @property type
	 * @since 0.7.0
	 */
	public var type: Type
		private set

	/**
	 * @property path
	 * @since 0.7.0
	 */
	public var path: String
		private set

	/**
	 * @property data
	 * @since 0.7.0
	 */
	public val data: String by lazy {
		this.read()
	}

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 * @hidden
	 */
	init {
		this.context = context
		this.type = type
		this.path = path
	}

	/**
	 * @method read
	 * @since 0.7.0
	 * @hidden
	 */
	private fun read(): String {

		try {

			if (this.path.startsWith("http://") ||
				this.path.startsWith("https://")) {
				return RemoteFileLoader().execute(this.path).get()
			}

			return this.context.assets.open(this.path).reader().use {
				it.readText()
			}

		} catch (e: IOException) {
			throw Exception("Cannot load source at location ${this.path}.")
		}
	}
}